### PR TITLE
Add a converter from mamba_ssm -> huggingface mamba

### LINKED ISSUE #29631: Conversion Script for Mamba checkpoints (`mamba_ssm` -> `transformers`)
### Feature request

Thanks very much for the Mamba support (#28094), this interoperability is fantastic!

I wanted to ask if there were any utility (doesn't have to be clean, just functional) for converting checkpoints provided for use in the `mamba_ssm` library into the format provided in `transformers`. 

This would be very helpful if it exists! Thanks 🤗 

### Motivation

I'd like to be able to convert novel trained mamba models from the `state-spaces/mamba` repo into HF transformers without rewriting a conversion script myself if need be.

### Your contribution

I could write a utility for this if none exists but would probably not have the bandwidth to upstream it.

### PR DESCRIPTION
**Pull request by byi8220**
# What does this PR do?
Fixes #29631

This PR creates a new command line script `convert_mamba_ssm_checkpoint_to_pytorch.py` which converts model checkpoints created by the [state-spaces/mamba](https://github.com/state-spaces/mamba/tree/main) repo into a Huggingface `MambaForCausalLM` model.

The intended usage of this script is:
```
python src/transformers/models/mamba/convert_mamba_ssm_checkpoint_to_pytorch.py \
--mamba_checkpoint_file=path/to/pytorch_model.bin \
--config_json_file=/path/to/ssm_config.json \
--output_dir=/path/to/out/dir
```

This script has a dependency on the `mamba_ssm` package.

### Testing

A validation pass is performed before exporting the model.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [X] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [X] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by amyeroberts** [comment]
Thanks for opening this PR @byi8220 ! Let us know when it's ready for review 🤗

**Comment by byi8220** [comment]
Thanks, I think it should be reviewable now?

File: tests/models/mamba/test_modeling_mamba.py [hunk 1]
```diff
@@ -489,3 +500,49 @@ def test_simple_generate_cuda_kernels_big(self, device):
         output_sentence = self.tokenizer.decode(output[0].tolist())
 
         self.assertEqual(output_sentence, expected_output)
+
+    @parameterized.expand([("state-spaces/mamba-130m",)])
```
Comment (amyeroberts) [inline tests/models/mamba/test_modeling_mamba.py:504]: I don't think it's necessary to add a test - we don't do this for other models. Instead we typically compare the original model's logits to our converted model's for a given output in the conversion script

**Review by amyeroberts — COMMENTED** [review]
Thanks for working on this!

Just a comment about having validation in the script rather than a new test

File: tests/models/mamba/test_modeling_mamba.py (hunk 1 above)
Comment (byi8220) [reply tests/models/mamba/test_modeling_mamba.py]: Makes sense. Removed the test and ran the command e2e and ran a spot test (https://gist.github.com/byi8220/9b44a5f6c6c2c7533704801478f1760a)

I was able to successfully convert their base 130m and 790m models without a problem, however my GPU setup runs out of memory trying for the 1.4b and 2.8b models.

File: src/transformers/models/mamba/convert_mamba_ssm_checkpoint_to_pytorch.py [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+import torch
+
+from transformers import AutoTokenizer, MambaConfig, MambaForCausalLM
+from transformers.utils import logging
+from transformers.utils.import_utils import is_mamba_ssm_available
+
+
+if is_mamba_ssm_available():
+    from mamba_ssm.models.config_mamba import MambaConfig as MambaConfig_ssm
```
Comment (amyeroberts) [inline src/transformers/models/mamba/convert_mamba_ssm_checkpoint_to_pytorch.py:30]: nit - camel casing for classes

```suggestion
    from mamba_ssm.models.config_mamba import MambaConfig as MambaConfigSSM
```

File: src/transformers/models/mamba/convert_mamba_ssm_checkpoint_to_pytorch.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+) -> None:
+    """Validate the converted model returns the same output as the original model."""
+    if not is_mamba_ssm_available():
+        raise ImportError(
+            "Calling validate_converted_model requires the mamba_ssm library to be installed. Please install it with `pip install mamba_ssm`."
+        )
+    if not torch.cuda.is_available():
+        raise ValueError(
+            "This script is to be run with a CUDA device, as the original mamba_ssm model does not support cpu."
+        )
```
Comment (amyeroberts) [inline src/transformers/models/mamba/convert_mamba_ssm_checkpoint_to_pytorch.py:87]: Let's move these into `convert_mamba_checkpoint_file_to_huggingface_model_file`. 

Each function doesn't need to handle imports, we can delegate this to a single "function" which calls others and is responsible for input / env validation

**Review by amyeroberts — APPROVED** [review]
Thanks for adding this! 

Looks good to me, just some small nits. cc @ArthurZucker for reference

File: src/transformers/models/mamba/convert_mamba_ssm_checkpoint_to_pytorch.py (hunk 2 above)
Comment (byi8220) [reply src/transformers/models/mamba/convert_mamba_ssm_checkpoint_to_pytorch.py]: Done

File: src/transformers/models/mamba/convert_mamba_ssm_checkpoint_to_pytorch.py (hunk 3 above)
Comment (byi8220) [reply src/transformers/models/mamba/convert_mamba_ssm_checkpoint_to_pytorch.py]: Done

**Comment by byi8220** [comment]
Addressed nits.

Since there's no unit test for this anymore just did one last sanity test. Works on my machine

1. My script used to spot test the conversion (https://gist.github.com/byi8220/9b44a5f6c6c2c7533704801478f1760a) passes for the 130m and 790m models. My GPU doesn't have enough memory to run the larger ones.
2. Downloading the weights and config of a [130m model online](https://huggingface.co/state-spaces/mamba-130m/tree/main) and running the command `python src/transformers/models/mamba/convert_mamba_ssm_checkpoint_to_pytorch.py --mamba_checkpoint_file=/tmp/scratch/pytorch_model.bin --config_json_file=/tmp/scratch/config.json --output_dir=/tmp/scratch/out` appears to function e2e.

**Comment by amyeroberts** [comment]
Awesome work - thanks again for adding this and for running some sanity checks!

**Pull request merged**